src/compositor.c

```diff
@@ -29,52 +29,120 @@
+ * mzrm         - Marius - Zico remote messaging (?). RPC used to talk between
+ *                main ARM core (named Marius on D6) and Zico.
+ * Ximr         - Render MIXer (?). Low level compositor on D6 and up.
+ * Ximr Context - Data structure representing Ximr configuration, sent via mzrm
+ *                using XimrExe() to Zico for render.
+ * Input layer  - Layer (described by MARV struct) to be included in mix.
+ * XOC          - Ximr Output Chunk. Buffer where final image is rendered.
+ * XCM          - Ximr Context Maker. "Userspace" tool to deal with Ximr setup.
+ *                Exists on D8 and up, from EOS R (M50 uses pure Ximr)
+ */
+
 #include "dryos.h"
 #include "bmp.h"
 #include "compositor.h"
 
-extern int uart_printf(const char * fmt, ...);
-
-#ifdef FEATURE_COMPOSITOR_XCM
+#ifdef CONFIG_COMPOSITOR_DEDICATED_LAYER
+/**
+ * refreshVrmsSurface (renamed to VMIX_TransferRectangleToVram on Digic X)
+ * is used by Canon code to render Input layers into XOC.
+ *
+ * On pure XIMR cameras it uses XIMR functions to setup Ximr Context
+ */
+extern void      refreshVrmsSurface();
+extern uint32_t  display_refresh_needed;
 
+// Store our Input Layer ID.
 int _rgb_vram_layer = 0;
 
-/*
- * Family of functions new to compositor.
- * Terminology:
- * Ximr - X image renderer (?). Compositor itself.
- * XCM  - Ximr Context Maker. "Userspace" tool to deal with Ximr setup.
- *        Exists on R, RP, DNE on m50, 200d.
- * XOC  - Ximr Output Chunk.
+/**
+ * Not sure if sync_caches() call is needed. It was when I was drawing
+ * over Canon buffers, but now when we have our own may be unnecessary.
  */
+void _compositor_force_redraw()
+{
+    display_refresh_needed = 1;
+    ml_refresh_display_needed = 1;
+    sync_caches();
+    //refreshVrmsSurface();
+}
 
-/*
- * XCM functions - for cameras that have it.
- * FEATURE_COMPOSITOR_XCM
- */
+void compositor_layer_clear()
+{
+    // abort if we draw over Canon GUI layer
+    if(_rgb_vram_layer == 0)
+        return;
 
-/*
- * Pure Ximr functions - for cameras without XCM
- * FEATURE_COMPOSITOR_XIMR
- */
+    bzero32(rgb_vram_info->bitmap_data, BMP_VRAM_SIZE*4);
+    _compositor_force_redraw();
+ }
 
-//TODO: Add Ximr support from 200D PoC
 
+struct MARV *_compositor_create_layer()
+{
+    // buffer for layer data
+    // looks like code expect it to be in uncacheable area
+    uint8_t* pBitmapData = UNCACHEABLE(malloc(BMP_VRAM_SIZE*4));
+    if(pBitmapData == NULL) return NULL;
+
+    // buffer for MARV structure
+    struct MARV* pNewLayer = malloc(sizeof(struct MARV));
+    if(pNewLayer == NULL) return NULL;
+
+    uint16_t bmp_w = BMP_W_PLUS - BMP_W_MINUS;
+    uint16_t bmp_h = BMP_H_PLUS - BMP_H_MINUS;
+
+    // fill MARV data
+    pNewLayer->signature    = 0x5652414D;  // MARV
+    pNewLayer->bitmap_data  = pBitmapData;
+    pNewLayer->opacity_data = 0x0;
+    // 0x5040100 is used by D78. Some early D6 use a different value
+    pNewLayer->flags        = 0x5040100;   // bitmask (?) for RGBA
```

Comment:
Better to make this a constant if you already know cams differ, easier for people to find and fix later

Reply:
Yes, probably should be a constant with a caveat - this is model dependent. At the moment I don't know any EOS codebase model that would use "old" IDs.

I'll create a constant in a follow up commit.

Reply:
Addressed in 5c4eea4cd3f5c502c692f772fc1f34541625f5f8